modules/compute/src/main/java/org/apache/ignite/compute/internal/IgniteComputeImpl.java

```diff
@@ -0,0 +1,302 @@
+        public void onDisappeared(ClusterNode member) {
+            executions.values().forEach(execution -> execution.notifyNodeDisappeared(member));
+        }
+    }
+
+    /** Handles execute commands and execution results. */
+    private class ComputeNetworkHandler implements NetworkMessageHandler {
+        /** {@inheritDoc} */
+        @Override
+        public void onReceived(NetworkMessage message, NetworkAddress senderAddr, String correlationId) {
+            short messageType = message.messageType();
+
+            switch (messageType) {
+                case ComputeMessagesType.EXECUTE_MESSAGE:
+                    assert message instanceof ExecuteMessage;
+
+                    handleExecuteCommand((ExecuteMessage) message, senderAddr);
+                    break;
+
+                case ComputeMessagesType.RESULT_MESSAGE:
+                    assert message instanceof ResultMessage;
+
+                    handleResult((ResultMessage) message, senderAddr);
+                    break;
+
+                default:
+                    throw new IllegalArgumentException("Unknown message type: " + messageType);
+            }
+        }
+    }
+
+    /**
+     * Handles an intermediate result sent by a node.
+     *
+     * @param message Message with the intermediate result.
+     * @param senderAddr Node's address.
+     */
+    private void handleResult(ResultMessage message, NetworkAddress senderAddr) {
+        UUID executionId = message.executionId();
+
+        @SuppressWarnings("rawtypes") Execution execution = executions.get(executionId);
+
+        ClusterNode clusterNode = clusterService.topologyService().getByAddress(senderAddr);
+
+        if (clusterNode == null) {
+            // Node left the topology, this will be handled by the TopologyHandler
+            return;
+        }
+
+        //noinspection unchecked
+        execution.onResultReceived(clusterNode.id(), message.result(), message.exception());
+    }
+
+    /**
+     * Handles the execution command sent by a node.
+     *
+     * @param msg Message with the execution command.
+     * @param senderAddr Node's address.
+     */
+    private void handleExecuteCommand(ExecuteMessage msg, NetworkAddress senderAddr) {
+        String className = msg.className();
+        Object argument = msg.argument();
+
+        executeOnExecutor(className, argument).whenComplete((result, throwable) -> {
+            if (throwable != null) {
+                throwable = throwable.getCause();
+
+                assert throwable instanceof ComputeException;
+            }
+
+            ResultMessage resultMessage = messagesFactory.resultMessage()
+                    .executionId(msg.executionId()).result(result).exception(throwable).build();
+
+            this.clusterService.messagingService().send(senderAddr, resultMessage, null);
+        });
+    }
+
+    private CompletableFuture<Object> executeOnExecutor(@NotNull String className, @NotNull Object argument) {
```

Comment:
Why is `argument` not null? This does not seem to be checked from `IgniteCompute#execute()` till this point.